### PR TITLE
plugin/rewrite: streamline the ResponseRule handling.

### PR DESCRIPTION
**Pull request by mandelsoft**
### 1. Why is this pull request needed and what does it do?

The functionality of a response rule is now completely encapsulated behind
a `ResponseRule` interface. This significantly simplifies the complete
processing flow, it enables more flexible response handling and it is possible
to eliminate lots of state flags, ifs and switches.

Based on the new flexibility the pull request also enables to support a
response name rewrite for all name rewrite types.
To be compatible, an explicit `answer auto` option is added for this.

Using this feature it is also possible now to rewrite a complete sub domain
hierarchy to a single domain name combined with a correct answer rewrite (coredns#2389).

The tests have been improved to test the new cases but also lots of older untested cases.

### 2. Which issues (if any) are related?
#2389, #2202

### 3. Which documentation changes (if any) need to be made?

The documentation of the rewrite plugin has been adapted to reflect the new answer option.

### 4. Does this introduce a backward incompatible change or deprecation?

No, the former behaviour completely remains. It just adds a new answer option for the rewrite plugin that can be used to influence the behaviour of all name rewrite types.

### CODE REVIEW

File: plugin/rewrite/README.md [hunk 1]
```diff
@@ -127,6 +128,28 @@ ftp.service.us-west-1.consul. 0    IN A    10.30.30.30
...
+```
+    rewrite stop {
+        name suffix .coredns.rocks .service.consul answer auto
+    }
+```
+
+Any occurrence of the rewritten question in the answer is mapped
+back to the original value before the rewrite.
+
+Please not that answers for rewrites of type `exact` are always rewritten.
```
Comment (chrisohaver) [inline plugin/rewrite/README.md]:
```suggestion
Please note that answers for rewrites of type `exact` are always rewritten.
```

File: plugin/rewrite/name.go [hunk 2]
```diff
@@ -58,62 +114,62 @@ const (
 
 // Rewrite rewrites the current request based upon exact match of the name
 // in the question section of the request.
```
Comment (miekg) [inline plugin/rewrite/name.go:116]: this comment is not true anymore, becasuse the rule is returned instead of executed?

File: plugin/rewrite/name.go [hunk 3]
```diff
@@ -58,62 +114,62 @@ const (
...
+		return nil, RewriteIgnored
 	}
 	s := rule.Replacement
 	for groupIndex, groupValue := range regexGroups {
 		groupIndexStr := "{" + strconv.Itoa(groupIndex) + "}"
 		s = strings.Replace(s, groupIndexStr, groupValue, -1)
 	}
 	state.Req.Question[0].Name = s
-	return RewriteDone
+	return rule.Response(state)
```
Comment (miekg) [inline plugin/rewrite/name.go]: why does this need to be returned? `rule.Response(state)` can also be called outside of this method

Comment (mandelsoft) [reply plugin/rewrite/name.go]: This is a dedicated implementation for the name rewrite types (the other types (can) do it differently), that need to dynamically handle the auto and non-auto case. I've separated this into this response function, because it is identical for all name types. `Response` is just an attribute, here. The interface function to be used from the outside is still the `Rewrite` method of a rule object, which returns a dedicated, possibly request specific response rule . Because of this feature to provide a request specific response rule, a `GetResponse()` method cannot be used for the rewrite rule type object. The rewrite object must have the full control depending on the rewrite request to provide a dedicated response rule.
Basically all the attributes should be internal(lower case), but they were public so so far, therefore I didn't changed this.

**Comment by mandelsoft** [comment]
I get this strange error in the kubernetes tests

```
go test -v ./...
go: github.com/coredns/caddy@v1.1.0: missing go.sum entry; to add it:
	go mod download github.com/coredns/caddy
```

Does anybody know how to get rid of it. I've done the download and the entry is already in the go.sum file.

**Comment by mandelsoft** [comment]
I get this strange error in the kubernetes tests

```
go test -v ./...
go: github.com/coredns/caddy@v1.1.0: missing go.sum entry; to add it:
	go mod download github.com/coredns/caddy
```

Does anybody know how to get rid of it? I've done the download and the entry is already in the go.sum file.

File: plugin/rewrite/name.go (hunk 3 above)
Comment (mandelsoft) [reply plugin/rewrite/name.go]: May be it is better to encapsulate the common behaviour of those name rules into a dedicated base struct used for the name rule types. I could provide this, It could also handle the `Mode()` function.

Comment (mandelsoft) [reply plugin/rewrite/name.go]: I've changed this now and introduced a common base part, which allowed to get rid of even more coding

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/rewrit..." ]
>@mandelsoft commented on this pull request.
>
>━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>
>In plugin/rewrite/name.go:
>
>>       }
>        s := rule.Replacement
>        for groupIndex, groupValue := range regexGroups {
>                groupIndexStr := "{" + strconv.Itoa(groupIndex) + "}"
>                s = strings.Replace(s, groupIndexStr, groupValue, -1)
>        }
>        state.Req.Question[0].Name = s
>-       return RewriteDone
>+       return rule.Response(state)
>
>I've changed this now and introduced a common base part, which allowed to get
>rid of even more coding

excellent. I'll take another look

File: plugin/rewrite/name.go (hunk 3 above)
Comment (miekg) [reply plugin/rewrite/name.go]: making members/functions private is OK w/ me. I think this was mostly a side effect of style. There should be nothing consuming this as an API

File: plugin/rewrite/name.go [hunk 4]
```diff
@@ -140,45 +193,56 @@ func newNameRule(nextAction string, args ...string) (Rule, error) {
...
+				return nil, fmt.Errorf("answer name mapping only for regexp rule")
+			}
+			if strings.ToLower(args[4]) != "name" {
+				return nil, fmt.Errorf("exceeded the number of arguments for a regex name rule")
+			}
+		}
 	}
 
 	if len(args) < 7 {
+		_ = auto
```
Comment (miekg) [inline plugin/rewrite/name.go]: why?

File: plugin/rewrite/name.go [hunk 5]
```diff
@@ -140,45 +193,56 @@ func newNameRule(nextAction string, args ...string) (Rule, error) {
...
 		switch matchType {
 		case ExactMatch:
-			rewriteAnswerFromPattern, err := isValidRegexPattern(rewriteQuestionTo, rewriteQuestionFrom)
+			_, err := isValidRegexPattern(rewriteQuestionTo, rewriteQuestionFrom)
 			if err != nil {
```
Comment (miekg) [inline plugin/rewrite/name.go:393]: this if can no be folded into the check, saving a line

File: plugin/rewrite/name_test.go [hunk 6]
```diff
@@ -75,13 +75,244 @@ func TestRewriteNamePrefixSuffix(t *testing.T) {
...
+}
+
 func TestNewNameRule(t *testing.T) {
 	tests := []struct {
 		next         string
 		args         []string
 		expectedFail bool
 	}{
-		{"stop", []string{"exact", "srv3.coredns.rocks", "srv4.coredns.rocks"}, false},
+		{"stop", []string{"exact", "rv3.coredns.rocks", "srv4.coredns.rocks"}, false},
```
Comment (miekg) [inline plugin/rewrite/name_test.go]: why this change?

File: plugin/rewrite/reverter.go [hunk 7]
```diff
@@ -23,15 +25,16 @@ type ResponseRule struct {
...
 	dns.ResponseWriter
 	originalQuestion dns.Question
-	ResponseRewrite  bool
 	ResponseRules    []ResponseRule
+	noSave           bool
 }
 
 // NewResponseReverter returns a pointer to a new ResponseReverter.
-func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg) *ResponseReverter {
+func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg, noSaveQuestion bool) *ResponseReverter {
```
Comment (miekg) [inline plugin/rewrite/reverter.go]:
```suggestion
func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg, restoreQuestion bool) *ResponseReverter {
```

File: plugin/rewrite/reverter.go [hunk 8]
```diff
@@ -23,15 +25,16 @@ type ResponseRule struct {
...
 	return &ResponseReverter{
 		ResponseWriter:   w,
 		originalQuestion: r.Question[0],
+		noSave:           noSaveQuestion,
```
Comment (miekg) [inline plugin/rewrite/reverter.go]:
```suggestion
		restore:           restoreQuestion,
```

File: plugin/rewrite/reverter.go [hunk 9]
```diff
@@ -40,42 +43,13 @@ func (r *ResponseReverter) WriteMsg(res1 *dns.Msg) error {
 	// Deep copy 'res' as to not (e.g). rewrite a message that's also stored in the cache.
 	res := res1.Copy()
 
-	res.Question[0] = r.originalQuestion
-	if r.ResponseRewrite {
+	if !r.noSave {
```
Comment (miekg) [inline plugin/rewrite/reverter.go]:
```suggestion
	if !r.restore {
```

File: plugin/rewrite/rewrite.go [hunk 10]
```diff
@@ -33,34 +33,29 @@ const (
 type Rewrite struct {
 	Next     plugin.Handler
 	Rules    []Rule
-	noRevert bool
+	noRevert bool // for test
```
Comment (miekg) [inline plugin/rewrite/rewrite.go]: oh members that _are_ only used for testing, that sounds iffy. Can this be done in another why. I would very much not have this done this way

File: plugin/rewrite/ttl.go [hunk 11]
```diff
@@ -9,81 +9,90 @@ import (
...
+
+func (r *ttlResponseRule) RewriteResponse(rr dns.RR) {
+	rr.Header().Ttl = r.TTL
+}
+
 type exactTTLRule struct {
 	NextAction string
 	From       string
-	ResponseRule
+	response   *ttlResponseRule
```
Comment (miekg) [inline plugin/rewrite/ttl.go]: why do these need to be pointers now? I think added a NoopRule can also remove the need to make `nil` the thing that conveys 'noop'. Would also make this PR much smaller, 'cause these changes can all be reverted

File: plugin/rewrite/name.go (hunk 4 above)
Comment (mandelsoft) [reply plugin/rewrite/name.go]: sorry , an artefact from testing, I'll remove it

File: plugin/rewrite/reverter.go (hunk 8 above)
Comment (mandelsoft) [reply plugin/rewrite/reverter.go]: Better world be `noRestore`, the default (false) should be the standard behavior, which is restoring the original question in the response handling. For testing purposes this will then be set to true.

File: plugin/rewrite/rewrite.go (hunk 10 above)
Comment (mandelsoft) [reply plugin/rewrite/rewrite.go]: The `noRevert` was already there, it has not been introduced by this PR. I just added another such field, and added the comment telling the reason for this field. I just found no other usage for the already existing field, besides various test cases.

File: plugin/rewrite/ttl.go (hunk 11 above)
Comment (mandelsoft) [reply plugin/rewrite/ttl.go]: I'll use the same pattern with a `ttlRuleBase` as for the name rules, which simplifies the complete rule implementations.

File: plugin/rewrite/name_test.go (hunk 6 above)
Comment (mandelsoft) [reply plugin/rewrite/name_test.go]: not intended, sorry, I'll revert to the original one

**Comment by mandelsoft** [comment]
@miekg I've introduced a base rule for all the ttl rules, also, with a common response handling. This simplifies the whole rewrite and response handling. Additionally I icorprated all your remarks, but instead of using the field name `Restore` I uses `noRestore` because the default behavior should be selected by the default field value (false).

I've added the field comment`// for test` because this field (`noRevert` was not introduced by this PR) was only used for test purposes, so far. The same is true for the `noRestore` field I added. If you see any usecase for package users we should make these field public. As far as they are package internal, they are only used for testing. 

If you don't like such testing fields in non-testing code, do you have a better idea? There could be a package local function enabling the test mode, but this would prevent parallel tests and the conscious decision for a dedicated test case to explicitly avoid those steps to check intermediate processing step results. Another possibility would be to officially introduce a policy struct, which controls the revert and restoration behaviour. Nevertheless it is basically only used for test purposes. 

If you think this is better, I can do it on short notice.

**Comment by mandelsoft** [comment]
Unfortunately the rewrite plugin has been changed in the meantime. Those changes are heavily conflicting, so I've to redo the changes.

**Comment by mandelsoft** [comment]
@miekg I've rebased and extended the auto rewrite by the additional value rewrite features. Therefore I aligned the value rewrite coding with the new restructuring of the rewrite plugin. Thereby a simplified and more flexible option parsing has been introduced to support all the variants. Additionally response rewrite rules are possible now for all name rewrites.

**Comment by mandelsoft** [comment]
@miekg Is there still something I can do?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/rewrit..." ]
>@miekg Is there still something I can do?

I'm sorry I didn't have any time to review this week (and probably also this weekend). So
hopefully Monday

File: .github/workflows/go.test.yml [hunk 12]
```diff
@@ -79,4 +79,4 @@ jobs:
       uses: actions/checkout@v2
 
     - name: Test Makefile
-      run: make -f Makefile.release build
+      run:  make DOCKER=bla GITHUB_ACCESS_TOKEN=bla -n build docker github-push docker-push -f Makefile.release
```
Comment (miekg) [inline .github/workflows/go.test.yml]: is this a rebase gone wrong?

File: man/coredns-rewrite.7 [hunk 13]
```diff
@@ -18,7 +18,7 @@ A simplified/easy-to-digest syntax for \fIrewrite\fP is...
 .RS
 
 .nf
-rewrite [continue|stop] FIELD [FROM TO|FROM TTL]
+rewrite [continue|stop] FIELD [TYPE] [(FROM TO)|TTL] [OPTIONS]
```
Comment (miekg) [inline man/coredns-rewrite.7]: revert any changes to this file, will be done automatically from a github workflow

File: man/coredns-sign.7 [hunk 14]
```diff
@@ -8,9 +8,9 @@
 .SH "DESCRIPTION"
```
Comment (miekg) [inline man/coredns-sign.7]: please revert all these changes

File: plugin/rewrite/name.go [hunk 15]
```diff
@@ -9,40 +9,117 @@ import (
...
+
+	"github.com/miekg/dns"
 )
 
-type exactNameRule struct {
-	NextAction string
-	From       string
-	To         string
-	ResponseRule
+////////////////////////////////////////////////////////////////////////////////
```
Comment (miekg) [inline plugin/rewrite/name.go]: please remove these `////` blocks, it clutters the godoc

File: plugin/rewrite/rewrite.go [hunk 16]
```diff
@@ -31,37 +31,29 @@ const (
 
 // Rewrite is a plugin to rewrite requests internally before being handled.
 type Rewrite struct {
-	Next     plugin.Handler
-	Rules    []Rule
-	noRevert bool
+	Next      plugin.Handler
+	Rules     []Rule
+	noRevert  bool // for test
```
Comment (miekg) [inline plugin/rewrite/rewrite.go]: could you add a TODO here, that we want to get rid of these 'used for testing members' in this struct.

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "Re: [coredns/coredns] plugin/rewrit..." ]
***@***.*** I've rebased and extended the auto rewrite by the additional value
>rewrite features. Therefore I aligned the value rewrite coding with the new
>restructuring of the rewrite plugin. Thereby a simplified and more flexible
>option parsing has been introduced to support all the variants. Additionally
>response rewrite rules are possible now for all name rewrites.

yep, I think this is very much a wanted clean up and improvement of the code, thanks for
you're continued work on it.

Also saw the rebase going a bit haywire you have no picked up a couple of other changes.

One real comment on the member in the Rewrite struct, just used for testing (add a TODO
there), but otherwise I see no reason to delay this any further.

Cheers!

File: .github/workflows/go.test.yml (hunk 12 above)
Comment (mandelsoft) [reply .github/workflows/go.test.yml]: I haven't touched the Makefile, but there is an additional change in the meantime. I just did a new rebase.

File: plugin/rewrite/rewrite.go (hunk 16 above)
Comment (mandelsoft) [reply plugin/rewrite/rewrite.go]: I could propose a general `RevertPolicy` instead, that can officially be set using the API.
Please give me a trigger to push that proposal. So far, I just added the TODO.

**Comment by mandelsoft** [comment]
@miekg I've replaced the revert options now by a RevertPolicy. This is a commit on top of the rest. So we can just discard it if you feel more comfortable with the TODO comment

**Comment by mandelsoft** [comment]
Hi @miekg , could we bring the PR somehow to an end?

**Review by miekg — APPROVED** [review]
